Tests/NIOPosixTests/DatagramChannelTests.swift

```diff
@@ -912,3 +906,11 @@ final class DatagramChannelTests: XCTestCase {
         testEcnAndPacketInfoReceive(address: "::1", vectorRead: true, vectorSend: true, receivePacketInfo: true)
     }
 }
+
+final class ConnectedDatagramChannelTests: DatagramChannelTests {
```

Comment:
I think this will stop working once we adopt test discovery on Linux, which strictly works structurally and cannot traverse into subclasses. Can we either use copy-paste or have two separate classes which both defer to a common implementation?

Reply:
Oh, that's disappointing. I didn't realise that.

`grpc-swift` makes use of subclasses, in that every test inherits from `GRPCTestCase` and there is further subclassing for shared tests IIRC. It also makes use of test discovery on Linux, given it's `LinuxMain.swift` consists only of:

```swift
fatalError("Use --enable-test-discovery to run tests on Linux")
```

What am I missing that means we can't do the same in NIO?

Reply:
We should double-check that all of grpc-swift's tests are actually running on Linux: my experiments suggest the answer is no.

Reply:
Looking into this:

- `GRPCTestCase` adds functionality but does not define tests, so that parent class is safe.
- `EchoTestCaseBase` adds functionality but does not define tests, so that parent class is safe.
- `FunctionalTestsInsecureTransport` defines tests, which is sketchy
- `GRPCInsecureInteroperabilityTests` defines tests, which is sketchy
- `ServerThrowingTests` defines tests, which is sketchy

Testing this locally I can confirm that the `FunctionalTestsInsecureTransport` subclasses `FunctionalTestsAnonymousClient` and `FunctionalTestsMutualAuthentication` do not run. The `GRPCInsecureInteroperabilityTests` subclass `GRPCSecureInteroperabilityTests` do not run. The `ServerThrowingTests` subclasses `ServerDelayedThrowingTests`, `ClientThrowingWhenServerReturningErrorTests`, `ServerErrorTransformingTests` do not run. This is a bug against grpc-swift.

Reply:
Filed as grpc/grpc-swift#1392.

Reply:
I've added tests for the connected support to the existing class for now. I'll resolve this thread and if it turns out that we want to have separate suite, I'll make sure I don't make the mistake of subclassing :)